ProjFS.Linux/PrjFSLib.Linux.Managed/VirtualizationInstance.cs

```diff
@@ -307,36 +288,40 @@ private int HandleNonProjEvent(ref Interop.ProjFS.Event ev, bool perm)
                 return 0;
             }
 
-            bool isDirectory = (ev.Mask & Interop.ProjFS.Constants.PROJFS_ONDIR) != 0;
+            bool isDirectory = (ev.Mask & ProjFS.Constants.PROJFS_ONDIR) != 0;
             string triggeringProcessName = GetProcCmdline(ev.Pid);
             byte[] providerId = new byte[PlaceholderIdLength];
             byte[] contentId = new byte[PlaceholderIdLength];
+            string relativePath = PtrToStringUTF8(ev.Path);
             Result result = Result.Success;
 
-            unsafe
+            if (!isDirectory)
             {
-                string relativePath = PtrToStringUTF8(ev.Path);
+                string currentRelativePath = relativePath;
 
-                if (!isDirectory)
+                // TODO(Linux): can other intermediate file ops race us here?
+                if (nt == NotificationType.FileRenamed)
                 {
-                    result = this.projfs.GetProjAttrs(
-                        relativePath,
-                        providerId,
-                        contentId);
+                    currentRelativePath = PtrToStringUTF8(ev.TargetPath);
                 }
 
-                if (result == Result.Success)
-                {
-                    result = this.OnNotifyOperation(
-                        commandId: 0,
-                        relativePath: relativePath,
-                        providerId: providerId,
-                        contentId: contentId,
-                        triggeringProcessId: ev.Pid,
-                        triggeringProcessName: triggeringProcessName,
-                        isDirectory: isDirectory,
-                        notificationType: nt);
-                }
+                result = this.projfs.GetProjAttrs(
+                    currentRelativePath,
+                    providerId,
+                    contentId);
+            }
+
+            if (result == Result.Success)
```

Comment:
Under what conditions will `GetProjAttrs` return false?

This is a bit unrelated to this PR but I wanted to mention that VFS4G needs notifications for all files inside of its root.  If `GetProjAttrs` returns false for newly created files inside the repo (i.e. those that it never projected), then VFS4G will be missing notifications it needs.

Reply:
Good question.  We should be returning notifications for all files, regardless of whether they have projection xattrs, because we treat that condition specially.

So `GetProjAttrs()` would only return a non-`Success` `Result` error code if something happened like the file was missing.

If an event is triggered for a file which lacks projection attributes (e.g., a file which VFS4G did not create), that wouldn't produce an error here, because we treat `ENOATTR` specially in libprojfs.  In this case the `providerId` and `contentId` byte arrays will be left unchanged; I believe C# allocates bytes as `0` so the arrays will just have all-zero values.  Does that seem OK?

Reply:
That makes sense, thanks!